File: templates/problems/01_arrays_hashing/04_group_anagrams/typescript/solution.ts
/**
 * Group strings that are anagrams using sorted-key hashing.
 * Time: O(n * k log k) where k is max string length, Space: O(n * k)
 */
export function solveGroupAnagrams(strs: string[]): string[][] {
    const groups = new Map<string, string[]>();
    for (const s of strs) {
        const key = new Array(26).fill(0);
        for (let i = 0; i < s.length; i++) {
            key[s.charCodeAt(i) - 'a'.charCodeAt(0)]++;
        }
        const keyStr = key.join(',');
        if (!groups.has(keyStr)) {
            groups.set(keyStr, []);
        }
        groups.get(keyStr)!.push(s);
    }
    return Array.from(groups.values());
}

File: templates/problems/01_arrays_hashing/04_group_anagrams/python/solution.py
def solve_group_anagrams(strs: list[str]) -> list[list[str]]:
    """Group strings that are anagrams using sorted-key hashing.

    Time: O(n * k log k) where k is max string length, Space: O(n * k)
    """
    groups: dict[tuple[int, ...], list[str]] = {}
    for s in strs:
        key = [0] * 26
        for ch in s:
            key[ord(ch) - ord('a')] += 1
        key_tuple = tuple(key)
        if key_tuple not in groups:
            groups[key_tuple] = []
        groups[key_tuple].append(s)
    return list(groups.values())
